### DISCUSSION TITLE
How to mock an async method in submodule using dependency_overrides?

### DISCUSSION THREAD

**Discussion by mauricioribeiro** [fastapi/fastapi#7059] [Questions] [answered]
First of all, I made a massive research and read the FastAPI doc before open this question, but I believe my case is a specific case:

Basically, I have two endpoints: one to create and another update a resource. In both of them I call an async method to download a Azure blob file by given SAS token in the body. So, in my unit tests I want to mock this async method then it will return a test file instead download it from Azure. It's kinda simple, but I'm facing some issues.

Here is my pseudo-code:

### main.py

```Python
# FastAPI imports
from azure import download_resource

app = FastAPI()

@app.post("/my-resource")
async def create_resource(body):
    #  validation before create
    return await create_or_update_resource(body, "create")

@app.put("/my-resource")
async def update_resource(body):
   #  validation before update
   return await create_or_update_resource(body, "update")

async def create_or_update_resource(body, action: str):
    # business rules validation
    await download_resource(body.sas_token)
    return {"messsage": action}
```

### azure.py 
```Python
# Azure library imports

async def download_resource(sas_token):
    # downloads azure file, blob validation, bla bla bla
   return file
```

### tests.py 
```Python
# FastAPI imports
from main import app
from azure import download_resource

client = TestClient(app)

async def mock_download_resource(sas_token):
    # gets test file to mock instead download from azure
    return testfile

app.dependency_overrides[download_resource] = mock_download_resource

class ResourceTest(TestCase):

    def test_create_resource(self):
        response = client.post('/my-resource', json={"sas_token": "test"})

        self.assertEqual(response.status_code, 200)
```

### Description

I know I suppose to add `Depends(download_resource)` parameter in my route methods (according to https://fastapi.tiangolo.com/advanced/testing-dependencies/), but if I put on it, FastAPI starts to ask sas_token in query parameter and responds 422.

Ideally, I could add `Depends(download_resource)` in my reused method `create_or_update_resource` (main.py) like this:
```Python
async def create_or_update_resource(body, action: str, injected_download_resource=Depends(download_resource)):
    # business rules validation
    await injected_download_resource(body.sas_token)
    return {"messsage": action}
```
But, if I do this, I receive an exception saying that `'Depends' object is not callable` (probably because it should be add in methods with `@router` decorator). So, how I supposed to do to implement a mock using my azure submodule?

### Environment

* OS: Windows:
* FastAPI Version: 0.61.1
* Python version: 3.8.6

**Comment by mauricioribeiro** [comment, marked as answer]
Just to you know, I solved this problem by return a method instead the value itself in `mock_download_resource` and taking `sas_token` off from method parameters. This way, I could add `Depends(download_resource)`  in my create/update methods and worked fine 👍 

Here is the piece of code I had to change:
```Python
async def mock_download_resource():
    async def mock(sas_token):
        return testfile
    return mock
```

As I said, the following modification was needed in route method as well:
```Python
...
@app.put("/my-resource")
async def update_resource(body, injected_download_resource=Depends(download_resource)):
          return await create_or_update_resource(body, "update", injected_download_resource)
...
```

**Comment by tiangolo** [comment]
Thanks for reporting back and closing the issue :+1:
